packages/remix-dev/config/flat-routes.ts

```diff
@@ -316,3 +320,15 @@ function getRouteMap(
 
   return routeMap;
 }
+
+let routeRegex =
+  /(([+][/\\][^/\\:?*]+)|[/\\]((_index)|([^/\\:?*])))\.(tsx?|jsx?|mdx?)$$/;
```

Comment:
This regex is too complex for me to understand at a glance. Could we break this down into simpler pieces somehow?

For example, this regex matches `/a.js`, but not `/abc.js` nor `a.js`. I'm not really sure if that's intentional or not.
It also seems to only match after the final `/` so `/abc/d.js` is only matched on `/d.js`. If we only want to match after the final path separator, I think it'd be better to use `path` utilities to get the base filename first and then match against regexes.

In any case, I think this deserves at least a comment explaining it, but probably warrants doing it a different way to keep things readable/understandable.

Reply:
Yes, the regex is complex. Even more so in my package as it handles more conventions. Unfortunately, JavaScript doesn't support the `x` flag (ignore whitespace) to make the regex easier to read.

Since routes can be determined as either a filename `a.b.c.tsx` or a folder `a.b.c./_index.tsx`, I don't think you can just get the base filename.

The filepath in `isRouteModuleFile` will be the full path after `routes` folder. So will never start with `/`.

To match `flat-files` convention, this is the regex: `^[^\/\\:?*]+\.(tsx?|jsx?|mdx?)$`

- Start of path `^` 
- Followed by one or more non-path characters `[^\/\\:?*]+`
- Ending with `.tsx`, etc `\.(tsx?|jsx?|mdx?)$`

<img width="742" alt="image" src="https://user-images.githubusercontent.com/47168/211588805-e9f10943-58b4-4039-9101-b537df3d28cf.png">

To match `flat-folders` convention, this is the regex: `[\/\\]_index\.(tsx?|jsx?|mdx?)$`

- Path ends with `/_index.tsx`
- So it should ignore colocated files like `users/component.tsx`

<img width="734" alt="image" src="https://user-images.githubusercontent.com/47168/211593213-c06f9f94-e2e2-442d-9c8e-a6f9f1bbf52e.png">

Put them together and you get: `((^[^\/\\:?*]+)|([\/\\]_index))\.(tsx?|jsx?|mdx?)$`

<img width="745" alt="image" src="https://user-images.githubusercontent.com/47168/211593478-65e37df8-6f08-408f-b46c-d3150ebc3af1.png">

Reply:
instead of using a regex would using `isIndexRoute` be sufficient for non flat files?

Reply:
Probably, considering the new core feature is a subset of the `remix-flat-rotues` package. You don't need to handle quite as many cases.

But if you end up expanding the naming conventions, you'll most likely need to use regex at some point.